### ISSUE TITLE
Master currently generates static

### ISSUE THREAD

**Issue by jyegerlehner** [ibab/tensorflow-wavenet#125]
I think I've found that the current master, which by default uses fast_generation, generates static. If fast_generation is disabled, then it seems to generate correctly. 

This conclusion is based on generating only two short samples. But I noticed the same behaviour in a development branch in which I had merged in master.

**Comment by lelayf** [comment]
I'm making a similar observation. Last good master that is working for me with fast gen is 4dafacb.
Also i am now shuffling samples and wondering if that could impact the training process.

**Comment by Zeta36** [comment]
**Yes!!** Fast generation is working badly.

I made the next experiment:

I modify the TextReader I made for the WaveNet text generation version (#117). specifically: I made a **ImageReader**. This reader just load (using Pillow: PIL) a Gif image, rescale it to 32x24 (pic.resize((32,24), Image.ANTIALIAS)) and finally get the 8bit grayscale channel (using convert('L')).

After all this, I push the (0-255) bytes from the image (grayscale) into a np vector (img = np.array(pic)). This is the result when I print it after reading a Gif image (32x24 matrix like this):
[[ 48  51  58  60  60  61  68  71  71  74  75  74  71  64  66  63  58  55
   53  58  56  55  54  60  62  63  65  70  73  70  63  46]
 [ 66  78  86  87  90  93  95 102 103  97  83  68  58  55  56  56  56  48
   44  46  47  50  66  85  91  91  93  95  99 100  91  62]
 [ 62  76  79  80  83  90  95  82  55  46  41  43  43  44  44  43  44  42
   39  35  33  37  45  54  65  79  91  91  92  94  90  61]
 [ 66  78  86  86  87  91  82  47  38  46  46  46  47  48  46  45  46  47
   45  46  49  44  39  44  47  57  73  89  97  99  92  62]
....
....
....
 [145 161 156 153 154 158 158 147 151 116 135 125 106 125 138 138 131 150
  145 129 160 134  90 180 219 208 203 202 220 223 214 186]
 [180 192 172 149 160 172 161 147 135 143 106 154 157 131 137 150 140 125
  121 172 187 116 152 204 196 200 203 211 219 217 210 180]]

I reshape this matrix later to one column with img = img.reshape(-1, 1)

Ok. Now I train the WaveNet model feeding with this 1D-image vector but using just a single image ("tp.gif") for all the training. The original (not scale) image is this (and once scaled its matrix is the one above): 
![tp](https://cloud.githubusercontent.com/assets/17341905/19157917/c2d192b8-8bde-11e6-8a09-ac43f2bfd4f8.gif)
For training I use the base parameters (of the master) and overfit the model so after 4k steps I get a near 0 loss (~0.3).

But before generating, in the generate.py I change two things:

1) I initialize the waveform in this way: waveform = [48] (insted of using a random int value as we do in the master branch).

2) I change now this line:     
> sample = np.random.choice(np.arange(quantization_channels), p=prediction)
by this:
> sample = np.argmax(prediction)

So, in this way I'm going to use always the best chance and not a random choice between the Softmax distribution (why don't we use argmax in the master by the way???) 

Well. After all this, I generate first using the fast generation and this is the result generated:
![sample_on](https://cloud.githubusercontent.com/assets/17341905/19157861/95517cb8-8bde-11e6-8a76-284369c8ff6f.gif)

But, and this is the key: if I don’t use the fast generation (all the rest the same) **I get the original image!!:**
![sample](https://cloud.githubusercontent.com/assets/17341905/19157893/b2097fa4-8bde-11e6-8255-8e2604a6b4b7.gif)

Summary:  Fast generation is generating a different output than the standard generation, and here is the empirical prove. Wrong matrix generated by the Fast generation way:

[[ 48  51  58  60  60  61  68  71  71  74  75  74  71  64  66  63  58  55
   53  58  56  55  54  60  62  63  65  70  73  70  63  46]
 [ 66  78  86  87  90  93  95 102 103  97  83  68  58  55  56  56  56  48
   44  46  47  50  66  85  91  91  93  95  99 100  91  62]
 [ 62  76  79  80  83  90  95  82  55  46  41  43  43  44  44  43  44  42
   39  47  66  37  45  91  47  86  91  83  86  90  66  68]
 [ 66  97  59 162 102  96  73 119  96  76  46 158 156 136 155  88 134  85
   80  38  99  44  81 115  62 154  80  70  56  66  83  86]
 [ 92  61  87  67  85  64  97 137  80  83  42 103  42 137  66 136  47  97
   75 165 161 167 120 144 100  82  73  98 196  93  62  47]
 [ 68  73 115  96 138  73  76 133  93 192  75 163 153  55 122 170 138 120
   85 136 154 145 147  78  38  86  43  86 105  94  68  71]
 [133  87 100 102  73  73  42 216  42  42  77 229  54  42  44 140 113 100
   96  47  50  46 147  38  47  42  50 213  79 219  72 184]
 [133  76  69  88  97  63  86  97  82  95 138 173  66 129  93 137 105 107
  125  35  41 144 133  42 149 171 125 223 118 217  44  94]
 [212  47 103  93  77  82  75  69  73 149  76  42  56  47 223  47  99 115
   82  96  72 149  88 160 129 110  77 192  62 122  72  75]
 [223 138  53  91 111  88  75 100  81  73  79  39  44 103  74  59  85  76
   96  78 229 216  98 242 121 214 120 129 108 147 103  63]
 [ 47 158  73 210 202 219  73  71  73 212 129 187  77  97  38 216 145 219
   80  71  49  78  78 136 152 232 145  99  85  39 103 115]
 [ 56  56  90  73  87  86  97  70  87 102 149  77 150 154 125 167 129 155
  232 137 109 140 223 229  71  54  75  85  71 141  92 189]
 [ 71  91 129 149  82 188 128  74 171 225 157  39  81  74 217 161  47 212
   83  76  88 222 202 147 144 212 114 158  76  66  85  85]
 [ 81 100  85  58  58  54 119  74  81  47  73 102  77  73  77  96  54  39
  230 219  73 201  46 143  96 133 103 232  98 220 179 191]
 [ 88  60  61  82  71  78 223 171  38 205 168 179  83 114  66 100  69 126
  157  60 219  72  76  33  76  68  33  55 170  43  64 103]
 [ 77 119  82  75 133 158 183 158  71 142  80  75 106 129  47  97 147  42
  170  78  81 103  81  96 114 230 118  88 140 159 174  78]
 [116 103  96  71  71 119  85 113 113 196 223 191  73  73  42  73 220 165
  127  75  75  30 137  86  75  79 128  75  47 220 205 158]
 [ 91 138  88 154 124  44 153  43 196  97  42  56  56  56  50  78  68  85
   63  83  63  73 142 154 121  42  40 145 158 147 174  98]
 [ 75 230 113 141 138 126 132  82  42  85 188 125  77 165  81  73 137  85
   82  98  77  63 163  72 173  88 216 168 121 161 170 232]
 [188 125 102  60 188  91  73 141 138  65  85  85  93  86 125  47  81  88
  138  77  76 128  85 138  49  73 152 120 129 177 223 152]
 [ 73 219  59 160 163  82  42 135 158 155 227 137 122  40 150 103  73 103
   43  73 102  71  47  93 173 128 171 125  72 125 170 217]
 [ 78  94 212 126 121 219  93  80  47  93 212  93  66  85  82  82  85  78
  142 124 161 143 216 168 242 228  59 220 140 160 145 169]
 [132  81  97 162 180 212  86  77  54 138  77 199  79  97  68  86  30  80
  222 128 188  47  73  95 120  38  73  73  82  73 155 196]
 [104 181  40 179 202  42 163  42  96 220 225 228 133 170 143  75 155 188
   42  47 129 143  73  66  75 133  41 103  47 140  73 143]]

And the perfectly matrix generated by the standard way (all the rest being the same). As you can see this matrix is the same as the original (once scaled). And that's because the overfit and the near 0 loss a get: 

[[ 48  51  58  60  60  61  68  71  71  74  75  74  71  64  66  63  58  55
   53  58  56  55  54  60  62  63  65  70  73  70  63  46]
 [ 66  78  86  87  90  93  95 102 103  97  83  68  58  55  56  56  56  48
   44  46  47  50  66  85  91  91  93  95  99 100  91  62]
 [ 62  76  79  80  83  90  95  82  55  46  41  43  43  44  44  43  44  42
   39  35  33  37  45  54  65  79  91  91  92  94  90  61]
 [ 66  78  86  86  87  91  82  47  38  46  46  46  47  48  46  45  46  47
   45  46  49  44  39  44  47  57  73  89  97  99  92  62]
 [ 66  79  89  92  93  77  47  40  42  44  41  43  42  40  41  47  48  48
   53  50  44  42  44  47  49  53  56  61  91 100  93  64]
 [ 66  78  90  90  87  59  40  38  35  42  60  79  82  90 115 114  85  85
  126 152 137  82  49  48  48  51  51  50  84  97  92  66]
 [ 66  82  91  91  80  53  39  49  86 110 149 188 208 215 213 196 180 194
  211 225 229 162  86  69  51  47  46  56  86  96  97  68]
 [ 65  81  93  91  76  47  35  68 144 163 182 216 238 242 242 228 220 232
  229 230 232 220 145  71  56  48  47  66  86  89  92  66]
 [ 66  76  89  88  77  50  30  75 142 161 174 204 231 235 232 224 222 242
  242 236 223 215 185  91  51  44  38  63  83  85  89  66]
 [ 69  81  87  83  79  68  37  73 120 120 131 153 181 210 213 202 191 173
  152 136 120 120 156 138  59  67  91  85  84  87  86  66]
 [ 62  73  85  83  77  89  81  66  72  65  71  71  77 108 147 140 103  76
   68  83 103 115 136 160  76 105 171 128  85  85  85  68]
 [ 63  68  80  82  77  75  96  85  99 105 102 103 103  96 144 175 146 136
  127 135 167 183 161 135 136 160 179 129  80  86  85  67]
 [ 63  70  83  81  75  70  82 110 119 146 158 169 166 119 160 227 170 161
  194 199 203 189 171 167 176 202 184  98  79  84  83  66]
 [ 61  71  85  78  74  72  67 100 127 133 157 168 140 128 184 233 216 170
  151 165 175 177 177 174 159 154 109  77  79  81  82  67]
 [ 60  72  84  78  72  71  70  67  98 119 126 113 111 119 145 169 158 165
  158 138 134 162 165 163 112  69  77  80  75  84  84  66]
 [ 62  74  81  73  68  70  70  66  73 107 111  97 104 104 112 124 141 153
  147 140 135 151 155 144  96  72  76  77  76  81  81  68]
 [ 64  74  82  69  70  68  68  67  63  78 104 107 108 109 114 135 143 154
  168 167 148 144 136 134 111  85  85  76  77  79  80  72]
 [ 61  74  80  71  69  66  67  66  65  69  99 100 108 119 115 115 124 139
  159 157 139 126 126 152  99  75 152  88  75  82  80  70]
 [ 60  75  76  69  67  68  68  65  62 106 111  88  92 104 122 135 137 138
  132 118 115 122 149 170 113 160 215 128  65  79  81  72]
 [ 59  73  74  68  65  62  62  69 100 143  89  82 105  93  85  80  82  91
  105 124 138 159 169 143 170 215 211 196 124  91  80  65]
 [ 59  69  69  67  80 102 133 149 164 160  87  72 118 120 111 118 134 142
  144 149 162 159 143 189 222 213 223 222 216 205 188 165]
 [ 62  92 113 135 155 165 168 158 147 126 119  72 113 132 123 125 141 147
  152 148 129 117 162 228 217 228 216 201 214 230 238 216]
 [145 161 156 153 154 158 158 147 151 116 135 125 106 125 138 138 131 150
  145 129 160 134  90 180 219 208 203 202 220 223 214 186]
 [180 192 172 149 160 172 161 147 135 143 106 154 157 131 137 150 140 125
  121 172 187 116 152 204 196 200 203 211 219 217 210 180]]

I have to mention that to get this generated matrix I have to reshape the one column vector that WaveNet generate by a bi-dimensional one using (making the inverse process that I do in the ImageRead): img = img.reshape(24, 32), and then:

new_img = Image.fromarray(img)
new_img = new_img.convert('RGB')
new_img.save(filename)

I'm going to upload all the ImageReader and the **WaveNet image version** (that is _similar_ to the **PixelCNN** but using just one channel) to my personal repository as I did with the Text version of the WaveNet (#117).

I **encourage** you to generate again using the trained model all you have from the WaveNet, but **turning off** the Fast generation and looking for some improvement in the generated file. I would try too changing 
> sample = np.random.choice(np.arange(quantization_channels), p=prediction)
by:
> sample = np.argmax(prediction)

and generating once again.
Maybe you can get in this way something better than so far.

I think that this experiment is the prove that the model used here is correct (since I could get a **perfect** memorization of the image using the WaveNet core and overfitting). If the model were wrong, I could have never use the WaveNet to recover and generate a **perfect** copy of the original image. And, in other hand, I've showed that the fast generation export a different output than the standard way (!!??).

Don't know whether the implementation is bad or if it is the theoretical idea the wrong, but I think somebody should review it, and maybe we should think also about using np.argmax(prediction) in the generation.

I hope this can help the team to improve in this project because the WaveNet model it's really interesting for me. It's capable of doing lots of things easily. 

It would be a good idea too create a generic Reader because there is a lot of things you can model with a vector of real values (text characters, 1D-images, amplitude sound waves, etc.).

Best regards,
Samu.

**Comment by lemonzi** [comment]
We already tested argmax, and it didn't work at all -- the most likely
value at each step is too affected by the global, time-independent prior in
the distribution, and is almost always 0. Using `choice` generates a time
series that reflects the probability distribution predicted by the network.

As for the fast generation, there were proper tests where it did work. The
DeepMind team even acknowledged using a similar construct, so the
theoretical idea is valid. I would say there is probably some bug in one of
the new features introduced recently that affects a code path only used by
fast generation.

The images were very clear by the way, no need to copy-paste raw matrices ;)

El dj., 6 oct. 2016 a les 11:37, Samuel Graván (<notifications@github.com>)
va escriure:

**Comment by lemonzi** [comment]
As for the PixelCNN, this only captures relationships between pixels in the same row (or same column, I don't remember which way it goes)  -- shouldn't it use 2D convolutions to capture relationships with the rest of the neighbors?

**Comment by Zeta36** [comment]
I have used a master as old as 09-29-2016.
And I pasted the full raw matrix because may help to the one who try to fix the problem ;)

Regards.

**Comment by ibab** [comment]
I don't see which change after 4dafacbc42024797bf022eca6ca4f87d8c898e52 could have caused this. Will look into this as soon as I can find some time.

**Comment by nakosung** [comment]
@lemonzi I think that we don't have conditioned generation yet, so argmax always returns same value. After local/global conditioning implemented we might stick to argmax for better result. (For an example, reinforcement agent using boltzmann distribution while learning, but its final build sticks to the argmax for optimal performance).

**Comment by Zeta36** [comment]
@ibab 

I think I found the problem with the fast generation. Problem probably was in a bug in this line in the _generator_dilation_layer() method:

conv_gate = output_gate + variables['gate_bias']

That it should be:

output_gate = output_gate + variables['gate_bias']

So, everyone that has used the bias parameter as True, has been generating noise for a long time. I test with the image generation (#125) and now all was fine with the fine fast generation (no different outputs between fast and standard generation).

Btw, looking at the model.py history, I can see this bug has been there since 09-27-2016 (the first release of the model.py file).

**Comment by lemonzi** [comment]
@Zeta3 Thanks a lot for the debugging!

@nakosung I think I'm missing something here. What does conditioning have to do with the argmax at the output?

**Comment by lelayf** [comment]
@ibab i think you should ignore my previous comment about 4dafacb working better than latest master, i think the dataset i'm using is the issue (lots of quasi-silence and i'm using threshold 0.01). I am not using biases. I'll do more investigation.

**Comment by nakosung** [comment]
@lemonzi With local conditioning argmax doesn't generate deterministic output (because its prior is ever-changing by local conditions). So after having local conditioning we can switch to argmax instead of some kind of distribution.

**Comment by Nyrt** [comment]
Oh cool, this looks like exactly the problem I was having.

**Comment by jyegerlehner** [comment]
I've confirmed that Zeta36 and nakosung's fix do resolve the problem I was seeing.
